File: server/src/module/performance.js
export function calculatePP(challengeInfo) {
  if (!challengeInfo) return null;

  const {
    difficulty,
    cnt_solvers,
    hitcount
  } = challengeInfo;

  try {

      const basePP = 10 * 1.5 ** difficulty; // Base PP
      const solveRate = cnt_solvers / hitcount * 100;  // 풀이 성공률 (%)
      const ppMultiplier1 = 10 / solveRate; // 10%를 기본 PP로 적용
      const ppMultiplier2 = 2 / Math.log10(Math.max(100, 100 + cnt_solvers)); // 풀이자 수가 많을 수록 패널티 (100명을 기본으로 둠)
      const ppMultiplier3 = 3 / Math.log10(Math.max(1000, 1000 + hitcount)); // 조회수가 많을 수록 패널티 (1000명을 기본으로 둠)

      const pp = basePP * ppMultiplier1 * ppMultiplier2 * ppMultiplier3;

      return pp;

  } catch (error) {
    console.error('PP 계산 중 오류:', error);
    return null;
  }
}

File: server/src/modules/performance.rs
use crate::modules::types::ChallengeInfo;
use libm::*;
use std::cmp::max;

pub fn calculate_pp(challenge_info: &ChallengeInfo) -> f64 {
    let base_pp = 10.0 * 1.5 * challenge_info.difficulty as f64;
    let solve_rate = challenge_info.cnt_solvers as f64 / challenge_info.hitcount as f64 * 100.0;
    let pp_multiplier1 = 10.0 / solve_rate;
    let pp_multiplier2 = 2.0 / log10(max(100, 100 + challenge_info.cnt_solvers) as f64);
    let pp_multiplier3 = 3.0 / log10(max(1000, 100 + challenge_info.hitcount) as f64);
    let pp = base_pp * pp_multiplier1 * pp_multiplier2 * pp_multiplier3;
    pp
}
